src/metaUtils.cxx

```diff
@@ -699,72 +702,56 @@ unsigned char * MET_PerformCompression(const unsigned char * source,
   // Choices are Z_BEST_SPEED,Z_BEST_COMPRESSION,Z_DEFAULT_COMPRESSION
   int compression_rate = Z_DEFAULT_COMPRESSION;
 
-  METAIO_STL::streamoff             buffer_size     = sourceSize;
-  unsigned char * input_buffer    = const_cast<unsigned char *>(source);
-  unsigned char * output_buffer   = new unsigned char[static_cast<size_t>(buffer_size)];
-
-  compressedData                  = new unsigned char[static_cast<size_t>(buffer_size)];
-
-  deflateInit(&z, compression_rate);
-
-  z.avail_in   = (uInt)buffer_size;
-  z.next_in    = input_buffer;
-  z.next_out   = output_buffer;
-  z.avail_out  = (uInt)buffer_size;
-
-  METAIO_STL::streamoff count;
-  METAIO_STL::streamoff j=0;
-  // Perform the compression
-  for ( ; ; )
+  METAIO_STL::streamoff buffer_in_size = sourceSize;
+  METAIO_STL::streamoff buffer_out_size = buffer_in_size;
+  METAIO_STL::streamoff max_chunk_size = std::numeric_limits<uInt>::max(); // could be smaller, but must not be bigger than UINT_MAX!
+  METAIO_STL::streamoff chunk_size = static_cast<uInt>(std::min(buffer_in_size, max_chunk_size));
+  unsigned char * input_buffer      = const_cast<unsigned char *>(source);
+  unsigned char * output_buffer     = new unsigned char[chunk_size];
+  unsigned char * compressed_data   = new unsigned char[buffer_out_size];
+
+  int ret = deflateInit(&z, compression_rate);
+  assert(ret == Z_OK);
+
+  METAIO_STL::streamoff cur_in_start = 0;
+  METAIO_STL::streamoff cur_out_start = 0;
+  int flush;
+  do
     {
-    if ( z.avail_in == 0 )
+    z.avail_in = static_cast<uInt>(std::min(buffer_in_size - cur_in_start, chunk_size));
+    z.next_in  = input_buffer + cur_in_start;
+    bool last_chunk = (cur_in_start + z.avail_in) >= buffer_in_size;
+    flush = last_chunk ? Z_FINISH : Z_NO_FLUSH;
+    cur_in_start += z.avail_in;
+    do
       {
-      deflate( &z, Z_FINISH );
-      count = buffer_size - z.avail_out;
-      if ( count )
+      z.avail_out = chunk_size;
+      z.next_out  = output_buffer;
+      ret = deflate(&z, flush);
+      assert(ret != Z_STREAM_ERROR);
```

Comment:
I thought before that assert was disabled in release builds everywhere, but on some more research I'm not so sure that applies for every platform. The assert calls should probably be removed or replaced by proper error handling.